include/hip/hcc_detail/functional_grid_launch.hpp

```diff
@@ -51,30 +51,53 @@ inline T round_up_to_next_multiple_nonnegative(T x, T y) {
     return tmp - tmp % y;
 }
 
-inline std::vector<std::uint8_t> make_kernarg() { return {}; }
-
-inline std::vector<std::uint8_t> make_kernarg(std::vector<std::uint8_t> kernarg) { return kernarg; }
-
-template <typename T>
-inline std::vector<std::uint8_t> make_kernarg(std::vector<uint8_t> kernarg, T x) {
-    kernarg.resize(round_up_to_next_multiple_nonnegative(kernarg.size(), alignof(T)) + sizeof(T));
-
-    new (kernarg.data() + kernarg.size() - sizeof(T)) T{std::move(x)};
-
+template <
+    std::size_t n,
+    typename... Ts,
+    typename std::enable_if<n == sizeof...(Ts)>::type* = nullptr>
+inline std::vector<std::uint8_t> make_kernarg(
+    std::vector<std::uint8_t> kernarg, const std::tuple<Ts...>&) {
     return kernarg;
 }
 
-template <typename T, typename... Ts>
-inline std::vector<std::uint8_t> make_kernarg(std::vector<std::uint8_t> kernarg, T x, Ts... xs) {
-    return make_kernarg(make_kernarg(std::move(kernarg), std::move(x)), std::move(xs)...);
+template <
+    std::size_t n,
+    typename... Ts,
+    typename std::enable_if<n != sizeof...(Ts)>::type* = nullptr>
+inline std::vector<std::uint8_t> make_kernarg(
+    std::vector<std::uint8_t> kernarg, const std::tuple<Ts...>& formals) {
+    using T = typename std::tuple_element<n, std::tuple<Ts...>>::type;
+
+    static_assert(
+        !std::is_reference<T>{},
+        "A __global__ function cannot have a reference as one of its "
+            "arguments.");
+    #if defined(HIP_STRICT)
+        static_assert(
+            std::is_trivially_copyable<T>{},
+            "Only TriviallyCopyable types can be arguments to a __global__ "
+                "function");
+    #endif
+
+    kernarg.resize(round_up_to_next_multiple_nonnegative(
+        kernarg.size(), alignof(T)) + sizeof(T));
+
+    new (kernarg.data() + kernarg.size() - sizeof(T)) T{std::get<n>(formals)};
+
+    return make_kernarg<n + 1>(std::move(kernarg), formals);
 }
 
-template <typename... Ts>
-inline std::vector<std::uint8_t> make_kernarg(Ts... xs) {
+template <typename... Formals, typename... Actuals>
+inline std::vector<std::uint8_t> make_kernarg(
+    void (*)(Formals...), std::tuple<Actuals...> actuals) {
+    static_assert(sizeof...(Formals) == sizeof...(Actuals),
+        "The count of formal arguments must match the count of actuals.");
+
+    std::tuple<Formals...> to_formals{std::move(actuals)};
```

Comment:
Is this move necessary or a way to erase the type of actuals in case of a mismatch?

Reply:
`move` doesn't do type erasure (or much at all really). `actuals` are sink parameters therefore it only makes sense to move from them as opposed to creating spurious copies.

Reply:
sorry I mis-read and I thought you added an extra move into to_formals such that make_kernarg would get instantiated with types of formals rather than actuals since they could be mismatched.  This looks fine

Reply:
`make_kernarg` takes the tuple of `Actuals` cast to `Formals` as its argument. The cast is part of constructing the `tuple`. Observe it is a `tuple<Formals...>` that is constructed from a pack of rvalue references to `Actuals`.

Reply:
I was referring to the subsequent call to make_kernarg which now takes a ref to the tuple of `Formals` so now the construction of the kernarg buffer is consistent with the kernel signature.